src/crypto/jh.c

```diff
@@ -213,16 +213,17 @@ static void E8(hashState *state)
 /*The compression function F8 */
 static void F8(hashState *state)
 {
-      uint64  i;
+      uint64_t* x = (uint64_t*)state->x;
+      const uint64_t* buf = (uint64*)state->buffer;
```

Comment:
This is technically undefined behavior according to the strict aliasing rule.

Reply:
Casting `uint8_t` buffer to `uint64_t` is technically UB, yes. But since we only XOR bytes here, it will not be affected by endianness. Compiling with `-fsanitize=undefined` doesn't find anything in this code. I will also add a proper data alignment for `buffer` to be sure we can access it as `uint64_t` without doing unaligned accesses.

Reply:
Also, even the old code did the same thing:
```
state->x[(8+i) >> 1][(8+i) & 1] ^= ((uint64*)state->buffer)[i];
```

Reply:
Even if it is correctly aligned, there is no guarantee by the language that cached, deferenced bytes will actually reflect the contents of `state->buffer` since the aliases are of non-compatible types. If we insert a `memcpy`, we can guarantee portability, and most compilers will optimize the superfluous instruction out anyways. I haven't looked into it, but the line `state->x[(8+i) >> 1][(8+i) & 1] ^= ((uint64*)state->buffer)[i];` could very well be one of the reasons that the old code broke on different optimization levels.

So instead of:
```
uint64_t* x = (uint64_t*)state->x;
const uint64_t* buf = (uint64*)state->buffer;

for (int i = 0; i < 8; ++i) x[i] ^= buf[i];

E8(state);

for (int i = 0; i < 8; ++i) x[i + 8] ^= buf[i];
```

I say we do:
```
uint64* x = state->x;

for (i = 0; i < 8; i++) 
{
    uint64 b;
    memcpy(&b, &state->buffer[i << 3], sizeof(b));
    x[i] ^= b;
}

E8(state);

for (i = 0; i < 8; i++) 
{
    uint64 b;
    memcpy(&b, &state->buffer[i << 3], sizeof(b));
    x[i + 8] ^= b;
}
```

It's a lot wordier, but it complies with the C++ standard.